app/models/user.rb

```diff
@@ -71,4 +73,10 @@ def sync_status_css_class
 
     css_class[sync_status]
   end
+
+  private
+
+  def add_city_id
+    self.city_id = Batch.find(batch_id).city_id unless batch_id.nil?
```

Comment:
On veut écraser le `city_id` qu'ils ont potentiellement défini ? 🤔 

Si non, on voudra peut-être faire soit
```suggestion
    self.city_id ||= Batch.find(batch_id).city_id unless batch_id.nil?
```
soit faire 
```ruby
  before_save :add_city_id, unless: :city_id?
```
au niveau de la déclaration du `before_save`.

Reply:
On peut aussi rework légèrement le find (un `batch_id` invalide raiserait une `ActiveRecord::RecordNotFound`) pour passer juste sur un comportement dégradé si c'est le cas:
```suggestion
    self.city_id ||= Batch.find_by(id: batch_id)&.city_id
```

Reply:
Le but de base c'etait de pas laisser le choix de la city a l'utilisateur ouai. T'as un batch ? Ta city c'ets ca point barre. Si les gens veulent changer ils peuvent toujours changer leur batch, par exemple pour un prof qui voudrait rejoindre l'equipe de son batch en cours, mais ca fausserait quand meme le top_contributor parce que le nombre d'alumni par city va pas changer.

D'après ce que Pilou me disait les seuls qui peuvent chosir leur city c'est les gens qui sont pas alumni genre ssaunier et c'est le cas parce que leur batch_id sont nil donc tant qu'ils entrent pas de batch dans leur settings leur ville choisit sera jamais ecrasé

Reply:
Je comprend pas ce que `Batch.find_by(id: batch_id)&.city_id` fait ni pourquoi tu passe d'un #find a un #find_by si c'est pour quand meme utiliser l'id 🤔

Reply:
D'ailleurs ca permettrait d'update toutes les villes des users existant en faisant un truc genre `User.all.each { _1.save }` dans la console

Reply:
@Aquaj I just revisited this and in the end I didn't change a thing.
Here's my thinking process :

- ```self.city_id ||= Batch.find(batch_id).city_id unless batch_id.nil?```
 I want the city to depend on the user's batch. If they had one and it wasn't the correct city then I'll consider this as cheating and they'll be redistributed to their correct city and they don't have a choice about that.
 That's why selecting the existing city_id isn't an option imo

-  `self.city_id Batch.find_by(id: batch_id)&.city_id`
The only case where the batch wouldn't be found is if the `batch_id` is `nil`.
The reason I don't do find_by is because I want to check if the `batch_id` is `nil` or not. In the user's settings, I disabled the batch's field so no one can edit theirs. IF a user doesn't not have a batch number then he can choose the city himself and the `before_save` won't run. That means that if the `before_save` runs the user has to have a valid `batch_id` hence why I don't need the find_by and the & imo